event_listener_test.go

```diff
@@ -173,3 +174,75 @@ func TestEventListener(t *testing.T) {
 		}
 	})
 }
+
+type writeDelayingFS struct {
+	vfs.FS
+	writesReleased chan struct{}
+}
+
+func (fs writeDelayingFS) Create(name string) (vfs.File, error) {
+	f, err := fs.FS.Create(name)
+	if err != nil {
+		return nil, err
+	}
+	if fs.writesReleased == nil {
+		return f, nil
+	}
+	return writeDelayingFile{f, fs.writesReleased}, nil
+}
+
+type writeDelayingFile struct {
+	vfs.File
+	writesReleased chan struct{}
+}
+
+func (f writeDelayingFile) Write(p []byte) (n int, err error) {
+	<-f.writesReleased
+	return f.File.Write(p)
+}
+
+// Verify `WriteStallBegin()` and `WriteStallEnd()` events. The setup logic
+// is a bit different so it is tested separately from the other events for now.
+func TestWriteStallEvents(t *testing.T) {
+	mem := vfs.NewMem()
+	customFS := writeDelayingFS{mem, nil /* writesReleased */}
+
+	stallBegan := make(chan struct{})
+	stallEnded := make(chan struct{})
+	listener := EventListener{
+		WriteStallBegin: func(info WriteStallBeginInfo) {
+			require.EqualValues(t, "memtable count limit reached", info.Reason)
+			close(stallBegan)
+		},
+		WriteStallEnd: func() {
+			close(stallEnded)
+		},
+	}
+	d, err := Open("db", &Options{
+		DisableWAL:                  true,
+		EventListener:               listener,
+		FS:                          customFS,
+		MemTableStopWritesThreshold: 2,
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// from now on delay writes until stall happens. Since WAL is disabled this results
+	// in delaying manifest writes, which blocks the flush thread. That causes memtables
+	// to accumulate, eventually resulting in a stall.
+	customFS.writesReleased = stallBegan
```

Comment:
This assignment probably came too late as the manifest is already created. I think the test coincidentally works because the second call to `AsyncFlush()` will typically happen before the background flush completes. I will try adding some sleeps to verify there is indeed a race here.

To fix it, we can have the `writeDelayingFile` hold a pointer to the channel in `writeDelayingFS` (and create all files as `writeDelayingFile`s), so we can propagate the change instantaneously, rather than applying it only to files created after this line.